### PR TITLE
Add and enforce checkstyle, lint, findbugs, and pmd

### PR DESCRIPTION
**Pull request by SUPERCILEX**
I noticed two things that lead to this PR:

1. We never did a full gradle check
2. We had a checkstyle, but it was never enforced

I started trying to get checkstyle to work and ended up adding all the common CI analysis tools/checks.

I went through and added checkstyle, lint, findbugs, and pmd. PMD is the only check that isn't enforced because it still needs to be configured to ignore tests. (It doesn't like the underscores in the test method names)

In the process, I fixed a ton of style and code issues.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -11,12 +11,45 @@ cache:
...
     - tools
-    - build-tools-25.0.0
+    - build-tools-25.0.1
     - android-25
-
-    # Extras
-    - extra-google-m2repository
+    - platform-tools
     - extra-android-m2repository
-script: ./gradlew clean :library:testAll :library:prepareArtifacts
```
Comment (SUPERCILEX) [inline .travis.yml]: What was `prepareArtifacts` for?

File: build.gradle [hunk 2]
```diff
@@ -8,7 +8,7 @@ buildscript {
     }
 
     dependencies {
-        classpath 'com.android.tools.build:gradle:2.2.2'
+        classpath 'com.android.tools.build:gradle:2.3.0-alpha1'
```
Comment (samtstern) [inline build.gradle]: What does 2.3 do for us?

File: .travis.yml (hunk 1 above)
Comment (samtstern) [inline .travis.yml]: That's the command I run before uploading to jCenter, I want to run it on each build to make sure that the build will be "releasable".

File: build.gradle (hunk 2 above)
Comment (SUPERCILEX) [inline build.gradle]: 2.3 added new lint features including the baseline and a whole [bunch of other stuff](http://tools.android.com/tech-docs/lint-in-studio-2-3).

File: .travis.yml (hunk 1 above)
Comment (SUPERCILEX) [inline .travis.yml]: Sounds good, I added it as a dependency of gradle's `check`.

**Comment by SUPERCILEX** [comment]
Will do!

**Comment by samtstern** [comment]
See: https://github.com/firebase/FirebaseUI-Android/pull/418#issuecomment-265209250

**Comment by SUPERCILEX** [comment]
@samtstern merged.

**Comment by SUPERCILEX** [comment]
This is awesome! The build failed because I forgot to delete an unused string resource.

**Comment by samtstern** [comment]
That is cool, also I love describing a build failure as "awesome" :+1:

**Comment by SUPERCILEX** [comment]
Yeah, "useful" would probably be a better word 😄

**Comment by SUPERCILEX** [comment]
That's one thing I wasn't sure about, do we want to enforce a max line length?

**Comment by SUPERCILEX** [comment]
@samstern, good to go! See my comment above though.

**Comment by samtstern** [comment]
@SUPERCILEX regarding max line length I don't think we need to enforce that, I generally try to keep my lines under 100 chars but I subscribe to the golang philosophy:

'Go has no line length limit. Don't worry about overflowing a punched card. If a line feels too long, wrap it and indent with an extra tab.'

**Comment by samstern** [comment]
Hey, I think you have the wrong guy. Are you confusing me with @samtstern?

**Comment by samtstern** [comment]
@samstern yes I think he was, cool name though :smile:

**Comment by samstern** [comment]
It's the BEST name!!

**Comment by SUPERCILEX** [comment]
Oops, sorry @samstern that was a typo 😄

**Pull request merged**